crates/sui-move-build/src/linters/self_transfer.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This analysis flags transfers of an object to tx_context::sender(). Such objects should be
+//! returned from the function instead
+
+use move_ir_types::location::*;
+
+use move_compiler::{
+    cfgir::{
+        absint::JoinResult,
+        ast::Program,
+        visitor::{
+            LocalState, SimpleAbsInt, SimpleAbsIntConstructor, SimpleDomain, SimpleExecutionContext,
+        },
+        CFGContext, MemberName,
+    },
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        Diagnostic, Diagnostics,
+    },
+    hlir::ast::{Command, Exp, LValue, Label, ModuleCall, Type, Type_, Var},
+};
+use move_symbol_pool::Symbol;
+use std::collections::BTreeMap;
+
+const TRANSFER_FUNCTIONS: &[(&str, &str, &str)] = &[
+    ("sui", "transfer", "public_transfer"),
+    ("sui", "transfer", "transfer"),
+];
+
+const SELF_TRANSFER_DIAG: DiagnosticInfo = custom(
+    "Lint ",
+    Severity::Warning,
+    /* category */ 1,
+    /* code */ 1,
+    "non-composable transfer to sender",
+);
+
+//**************************************************************************************************
+// types
+//**************************************************************************************************
+
+pub struct SelfTransferVerifier;
+
+pub struct SelfTransferVerifierAI {
+    fn_name: Symbol,
+    fn_ret_loc: Loc,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Default)]
+pub enum Value {
+    /// an address read from tx_context:sender()
+    SenderAddress,
```

Comment:
If we add a `loc`, we can point to where the `sender()` call was, which might be helpful for more complicated code

Reply:
Good idea!